### ISSUE TITLE
AffineExpr::simplifyAdd bug + fix

### ISSUE THREAD

**Issue by bondhugula** [tensorflow/mlir#82]
Currently, the affine expr, d0 - (d0 + 1), for example, is simplified to 0 in AffineExpr::simplifyAdd. (Any d0 - c * (d0 <binOp> c) is being simplified to d0 mod c due to a missing conditional.)

func @foo(%M : index) -> index {
  %v = affine.apply (d0) -> (d0 -(d0 + 1)) (%M)
  return %v : index
}

$ mlir-opt test.mlir
#map0 = (d0) -> (0)

module {
  func @foo(%arg0: index) -> index {
    %0 = affine.apply #map0(%arg0)
    return %0 : index
  }
}

This is fixed by changing the following:

https://github.com/tensorflow/mlir/blob/e294a297da8f6fb21f29abe31bdc6997fc6e5650/lib/IR/AffineExpr.cpp#L344

to

```
if (!lrBinOpExpr || lrBinOpExpr.getKind() != AffineExprKind::FloorDiv)
```

**Comment by jpienaar** [comment]
Could you create a PR instead of an issue if you are proposing a fix? Thanks

**Comment by antiagainst** [comment]
This should be fixed by 09db8c1 now. Closing.
